src/gevent/pool.py

```diff
@@ -714,13 +723,31 @@ def free_count(self):
             return 1
         return max(0, self.size - len(self))
 
-    def add(self, greenlet):
+    def add(self, greenlet, blocking=True, timeout=None):
         """
-        Begin tracking the given greenlet, blocking until space is available.
+        Begin tracking the given greenlet, possibly blocking until space is
+        available.
+
+        :keyword bool blocking: If True (the default), this function will block
+        until the pool has space or a timeout occurs. If False, this function
+        will immediately raise a Timeout if the pool is currently full.
+        :keyword float timeout: The maximum number of seconds this method will
+        block, if ``blocking`` is True. (Ignored if ``blocking`` is False.)
+
+        Raises ``Timeout`` on timeout, or `PoolFull` if ``blocking`` is False and
+        the pool is full.
 
         .. seealso:: :meth:`Group.add`
+
+        .. versionchanged:: 1.3.0
+            Added the ``blocking`` and ```timeout`` parameters.
         """
-        self._semaphore.acquire()
+        if not self._semaphore.acquire(blocking=blocking, timeout=timeout):
+            # We failed to acquire the semaphore.
+            # If blocking was True, then there was a timeout. If blocking was
+            # False, then there was no capacity.
+            raise Timeout() if blocking else PoolFull()
```

Comment:
This should simply be `raise PoolFull()` (with corresponding test and docstring changes). I don't think we should be raising `Timeout` here. Not only does it make the API more complicated (a caller has to catch two kinds of exceptions), it breaks timeout handling completely. Because the caller didn't create the timeout object, the usual way of telling if the timeout that elapsed was yours or someone else's is broken:

```python
def do_something(blocking, timeout):
   greenlet = ...
   try:
     pool.add(greenlet, blocking, timeout)
   except PoolFull:
     # pool was full
     greenlet.kill()
  except Timeout:
     # UH-OH! We don't know if this is because the pool was full,
     # or because our caller had established a timeout of its own
```

Reply:
Hmm, I was trying to preserve the ability to distinguish timeouts from full pools, but after reading your comment and thinking it over, I agree that this is unnecessary - the caller, by virtue of choosing what to specify for `blocking` and `timeout` in the first place - already has enough information to determine whether a raise `PoolFull` was due to a timeout or a nonblocking full pool.

Will make the changes to code/docs/tests.